File: TaeBeomShin/5. 이진탐색/K번째_수.py
n=int(input())
k=int(input())

start,end,ans=1,k,0
while start<=end:
    mid=(start+end)//2
    cnt=0

    for i in range(1,n+1):
        cnt+=min(mid//i,n)
    
    if(cnt<k):
        start=mid+1
    else:
        ans=mid
        end=mid-1

print(ans)

File: jifrozen/이진탐색/k번째수.java
package 이진탐색;
import java.util.*;

public class k번째수 {
	public static void main(String[] args){
		Scanner sc=new Scanner(System.in);

		int n=sc.nextInt();
		int k=sc.nextInt();

		int start=1;
		int end=k;
		int result=0;

		while(start<=end){
			int mid=(end+start)/2;
			int cnt=0;
			for(int i=1;i<=n;i++){
				cnt+=Math.min(n,mid/i);
			}

			if(cnt>=k){
				result=mid;
				end=mid-1;
			}else{
				start=mid+1;
			}
		}

		System.out.print(result);

	}
}
